### PR TITLE
Add Extended messages data to binary log

### LINKED ISSUE #9091: [BinFmt] Support extended messages in binary log and binary viewer
#8917 will introduce new internal EventArgs `Extended*EventArgs`
In context of PR such events are stored in binary log as existing corresponding non extended `*EventArgs`.

As there is certain value in showing exteneded event data in binary log we need to:
- [x] create new binary log events
- [x] de/ser those
- [x] bump up [`FileFormat`](https://github.com/dotnet/msbuild/blob/main/src/Build/Logging/BinaryLogger/BinaryLogger.cs#L63)
- [x] facilitate updating binary log viewer to support those

### PR DESCRIPTION
**Pull request by rokonec**
Fixes #9091

### Context
We have introducing new extended event args (error, warn, message, custom) so time ago. 
This PR is about logging extended data into binary log and will serve for modifying bin log viewer as well.

### Changes Made
Add `BuildEventArgsFieldFlags.Extended` and implement related code.

### Testing
Unit tests

### CODE REVIEW

**Comment by rokonec** [comment]
@KirillOsenkov I suppose we shall 1st release forward compatible log viewer. We can do it in two phases:
1) make it read new file version
2) UI changes to somehow show extended data

While I can easily do 1 for 2 I might need some help.
We expect that after C++ teams implement extended error logs people might start to need it frequently.

**Comment by KirillOsenkov** [comment]
Do we have a design document or a wiki page outlining the problem space, the motivation and the proposed approach? What do we know about the existing usage of the "exotic" event args? Do we have a list of teams/scenarios depending on these?

As we discover scenarios and knowledge, let's add to the doc. I'm sure we'll break someone, so would be nice to point them to the doc. I've always felt uncertainty about the exotic event args - they're ancient and it's unclear who if anyone is still using them and for what. Would also help to know these scenarios so we can test and validate.

**Comment by rokonec** [comment]
@KirillOsenkov 

Research of usage custom event arguments was concluded in:
https://github.com/dotnet/msbuild/issues/8825

Custom event arguments has been deprecated for dotnet core msbuild in favor of new messages containing general purpose transparent payload. See https://github.com/dotnet/msbuild/pull/8917

I would not care too much about binlog supporting rare needs for custom events args.
However, for close future, C++ team is improving VS with enhanced structured errors and warnings info json format called SARIF. These structured data comes from C++ compiler.

We believe, that there can be substantial value in capturing and showing such info in log viewer. And this PR is the part of such effort.

**Comment by KirillOsenkov** [comment]
Sounds good. I missing a ton of context, but superficially everything seems fine.

See if you can utilize BuildEventArgsFieldFlags in any way:
https://source.dot.net/#Microsoft.Build/Logging/BinaryLogger/BuildEventArgsFieldFlags.cs,58948e4bf96933c9,references

**Comment by rokonec** [comment]
@KirillOsenkov I have totally missed `BuildEventArgsFieldFlags`. Have just simplified and rewritten it. Good catch. Thx

**Comment by rokonec** [comment]
It is ready for review. I just left it in Draft so it is not merged before we release compatible log viewer.

File: src/Build/Logging/BinaryLogger/ExtendedDataFields.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable disable
```
Comment (JanKrivanek) [inline src/Build/Logging/BinaryLogger/ExtendedDataFields.cs:4]: Can we define it with nullables?
Then we e.g. could use `ReadString` vs `ReadOptionalString` for the `ExtendedType`

**Review by JanKrivanek — APPROVED** [review]
Looks good!

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs [hunk 2]
```diff
@@ -672,7 +697,7 @@ private void WriteMessageFields(BuildMessageEventArgs e, bool writeMessage = tru
             }
         }
 
-        private void WriteArguments(object[] arguments)
+            private void WriteArguments(object[] arguments)
```
Comment (KirillOsenkov) [inline src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:700]: Adding extra unnecessary whitespace

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs [hunk 3]
```diff
@@ -465,6 +484,7 @@ private void Write(BuildMessageEventArgs e)
             }
         }
 
+
```
Comment (KirillOsenkov) [inline src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:487]: No need to add an extra empty line here

**Comment by KirillOsenkov** [comment]
Looks OK as far as I can see

File: src/Build/Logging/BinaryLogger/ExtendedDataFields.cs (hunk 1 above)
Comment (rokonec) [reply src/Build/Logging/BinaryLogger/ExtendedDataFields.cs]: Also changed to immutable (removed property setters - what were I thinking :-)

**Review by YuliiaKovalova — APPROVED** [review]
